ref/vk/vk_devmem.c

```diff
@@ -66,138 +197,219 @@ static int allocateDeviceMemory(VkMemoryRequirements req, int type_index, VkMemo
+ 	REGISTER_STATS_METRIC( stats->current.allocated,         current_allocated##usage_suffix,         g_devmem.stats[usage_suffix].current.allocated,         kSpeedsMetricBytes ); \
+ 	REGISTER_STATS_METRIC( stats->current.align_holes,       current_align_holes##usage_suffix,       g_devmem.stats[usage_suffix].current.align_holes,       kSpeedsMetricCount ); \
+ 	REGISTER_STATS_METRIC( stats->current.align_holes_size,  current_align_holes_size##usage_suffix,  g_devmem.stats[usage_suffix].current.align_holes_size,  kSpeedsMetricBytes ); \
+ 	REGISTER_STATS_METRIC( stats->peak.allocations,          peak_allocations##usage_suffix,          g_devmem.stats[usage_suffix].peak.allocations,          kSpeedsMetricCount ); \
+ 	REGISTER_STATS_METRIC( stats->peak.allocated,            peak_allocated##usage_suffix,            g_devmem.stats[usage_suffix].peak.allocated,            kSpeedsMetricBytes ); \
+ 	REGISTER_STATS_METRIC( stats->peak.align_holes,          peak_align_holes##usage_suffix,          g_devmem.stats[usage_suffix].peak.align_holes,          kSpeedsMetricCount ); \
+ 	REGISTER_STATS_METRIC( stats->peak.align_holes_size,     peak_align_holes_size##usage_suffix,     g_devmem.stats[usage_suffix].peak.align_holes_size,     kSpeedsMetricBytes ); \
+ 	REGISTER_STATS_METRIC( stats->peak.align_hole_size,      peak_align_hole_size##usage_suffix,      g_devmem.stats[usage_suffix].peak.align_hole_size,      kSpeedsMetricBytes ); \
 }
 
 qboolean VK_DevMemInit( void ) {
-	g_vk_devmem.verbose = !!gEngine.Sys_CheckParm("-vkdebugmem");
+	g_devmem.verbose = gEngine.Sys_CheckParm( "-vkdebugmem" );
+
+	// Register standalone metrics.
+	R_SPEEDS_METRIC( g_devmem.alloc_slots_count, "allocated_slots", kSpeedsMetricCount );
+	R_SPEEDS_METRIC( g_devmem.device_allocated, "device_allocated", kSpeedsMetricBytes );
+	
+	// Register stats metrics for each usage type.
+	REGISTER_STATS_METRICS( VK_DEVMEM_USAGE_TYPE_ALL,    _ALL );
+	REGISTER_STATS_METRICS( VK_DEVMEM_USAGE_TYPE_BUFFER, _BUFFER );
+	REGISTER_STATS_METRICS( VK_DEVMEM_USAGE_TYPE_IMAGE,  _IMAGE );
+	
 	return true;
 }
 
 void VK_DevMemDestroy( void ) {
-	for (int i = 0; i < g_vk_devmem.num_allocs; ++i) {
-		const vk_device_memory_t *const device_memory = g_vk_devmem.allocs + i;
-		ASSERT(device_memory->refcount == 0);
+	for ( int slot_index = 0; slot_index < g_devmem.alloc_slots_count; slot_index += 1 ) {
+		const vk_device_memory_slot_t *const slot = g_devmem.alloc_slots + slot_index;
+		ASSERT( slot->refcount == 0 );
 
 		// TODO check that everything has been freed
-		aloPoolDestroy(device_memory->allocator);
+		aloPoolDestroy( slot->allocator );
+
+		if ( slot->mapped )
+			vkUnmapMemory( vk_core.device, slot->device_memory );
+
+		vkFreeMemory( vk_core.device, slot->device_memory, NULL );
+	}
+
+	g_devmem.alloc_slots_count = 0;
+}
 
-		if (device_memory->map)
-			vkUnmapMemory(vk_core.device, device_memory->device_memory);
+const char *VK_DevMemUsageTypeString( vk_devmem_usage_type_t type ) {
```

Comment:
Не вижу пока, зачем это нужно было бы показывать наружу.
```suggestion
static const char *VK_DevMemUsageTypeString( vk_devmem_usage_type_t type ) {
```

Reply:
Спрятал.